gfx_direct3d11.cpp

```diff
@@ -0,0 +1,976 @@
+        nullptr,
+        d3d.context.GetAddressOf()));
+
+#if DEBUG_D3D
+    ThrowIfFailed(d3d.device->QueryInterface(__uuidof(ID3D11Debug), (void **) (d3d.debug.GetAddressOf())));
+#endif
+
+    // Create views
+
+    create_render_target_views(DESIRED_SCREEN_WIDTH, DESIRED_SCREEN_HEIGHT);
+
+    // Create main vertex buffer
+
+    D3D11_BUFFER_DESC vertex_buffer_desc;
+    ZeroMemory(&vertex_buffer_desc, sizeof(D3D11_BUFFER_DESC));
+
+    vertex_buffer_desc.Usage = D3D11_USAGE_DYNAMIC;
+    vertex_buffer_desc.ByteWidth = 256 * 26 * 3 * sizeof(float); // Same as buf_vbo size in gfx_pc
+    vertex_buffer_desc.BindFlags = D3D11_BIND_VERTEX_BUFFER;
+    vertex_buffer_desc.CPUAccessFlags = D3D11_CPU_ACCESS_WRITE;
+    vertex_buffer_desc.MiscFlags = 0;
+
+    ThrowIfFailed(d3d.device->CreateBuffer(&vertex_buffer_desc, NULL, d3d.vertex_buffer.GetAddressOf()));
+
+    // Create constant buffer
+
+    D3D11_BUFFER_DESC constant_buffer_desc;
+    ZeroMemory(&constant_buffer_desc, sizeof(D3D11_BUFFER_DESC));
+
+    constant_buffer_desc.Usage = D3D11_USAGE_DYNAMIC;
+    constant_buffer_desc.ByteWidth = sizeof(PerFrameCB);
+    constant_buffer_desc.BindFlags = D3D11_BIND_CONSTANT_BUFFER;
+    constant_buffer_desc.CPUAccessFlags = D3D11_CPU_ACCESS_WRITE;
+    constant_buffer_desc.MiscFlags = 0;
+
+    ThrowIfFailed(d3d.device->CreateBuffer(&constant_buffer_desc, NULL, d3d.per_frame_cb.GetAddressOf()));
+
+    // Show the window
+
+    QueryPerformanceFrequency(&frequency);
+    QueryPerformanceCounter(&last_time);
+
+    ShowWindow(h_wnd, SW_SHOW);
+    UpdateWindow(h_wnd);
+}
+
+static void gfx_d3d11_dxgi_main_loop(void (*run_one_game_iter)(void)) {
+    MSG msg = { 0 };
+
+    if (PeekMessage(&msg, NULL, 0, 0, PM_REMOVE)) {
+        TranslateMessage(&msg);
+        DispatchMessage(&msg);
+    } else {
+        if (IsIconic(h_wnd)) {
+            Sleep(1);
+            return;
+        }
+
+        LARGE_INTEGER current_time, elapsed_time_microseconds;
+        QueryPerformanceCounter(&current_time);
+        elapsed_time_microseconds.QuadPart = current_time.QuadPart - last_time.QuadPart;
+        elapsed_time_microseconds.QuadPart *= 1000000;
+        elapsed_time_microseconds.QuadPart /= frequency.QuadPart;
+
+        if (elapsed_time_microseconds.QuadPart < 33333.33333333) {
+            return;
+        }
+
+        last_time = current_time;
+
+        d3d.run_one_game_iter = run_one_game_iter;
+        if (d3d.run_one_game_iter != nullptr) {
+            d3d.run_one_game_iter();
+        }
+    }
+}
+
+static void append_str(char *buf, size_t *len, const char *str) {
+    while (*str != '\0') buf[(*len)++] = *str++;
+}
+
+static void append_line(char *buf, size_t *len, const char *str) {
+    while (*str != '\0') buf[(*len)++] = *str++;
+    buf[(*len)++] = '\r';
+    buf[(*len)++] = '\n';
+}
+
+static const char *shader_item_to_str(uint32_t item, bool with_alpha, bool only_alpha, bool inputs_have_alpha, bool hint_single_element) {
```

Comment:
Would it be possible to put common code for both d3d12 and d3d11 in a common place to avoid duplications?